### DISCUSSION TITLE
Template for downloading CSV of connections between devices and/or patch panels

### DISCUSSION THREAD

**Discussion by Prestwickuk** [netbox-community/netbox#17386] [Help Wanted!] [answered]
Hi there team, 

I've got a site which we administer through netbox which is relatively complex in terms of cable paths from device A to device B. Usually, there are two devices linked by 4-5 different cables across four different patch panels. 

What I wanted was to be able to download a cable route from device A to device B with each hop clearly marked in CSV format. Basically, I want the cable trace feature in CSV format for all connections originating in a specific rack?

Anyway, how best would I be able to go about this as its not immediately clear within netbox how I can ask it to do that for me. 

Thanks in advance!

**Comment by candlerb** [comment, marked as answer]
Most flexible would be to query from the REST API and write your own code to format however you like.

If you want to do this within Netbox then you could create an [export template](https://netboxlabs.com/docs/netbox/en/stable/customization/export-templates/). The method `trace()` on an interface gives you a list of the path elements. You can exercise this in [nbshell](https://netboxlabs.com/docs/netbox/en/stable/administration/netbox-shell/).

```
# /opt/netbox/venv/bin/python /opt/netbox/netbox/manage.py nbshell
...
>>> i = Interface.objects.get(name="ether8",device__name="gw2")
>>> i
<Interface: ether8>
>>> i.trace()
[([<Interface: ether8>], [<Cable: #51>], [<FrontPort: F3>]), ([<RearPort: R3>], [<Cable: #54>], [<Interface: eno1>])]
```

Note that you get a list of three-tuples here, each element of which of which has a *list* of terminations or cables (one cable can connect to multiple components at each end), so it's not exactly ideal for CSV representation. You'd need to do some [jinja2](https://jinja.palletsprojects.com/en/3.1.x/templates/) munging to make it some sort of acceptable format, and that's pretty ugly.

But a starting point would be something like this (briefly tested): this is an export template on "Interface", so you'd have to go to Interfaces and filter to the interface(s) of interest.

```
{%- for interface in queryset %}
# Cable trace for {{ interface }} on {{ interface.device.name }}
from,via,to
{% for cfrom, cvia, cto in interface.trace() %}
{%- for x in cfrom %}{{ x.name }}@{{ x.device.name }} {%endfor%},{% for x in cvia %}{{ x }} {%endfor%},{% for x in cto %}{{ x.name 
 }}@{{x.device.name }} {%endfor%}
{% endfor %}
--------
{% endfor %}
```

Sample output:

```
# Cable trace for ether8 on gw2
from,via,to
ether8@gw2 ,#51 ,F3@pp1
R3@pp1 ,#54 ,eno1@brian-kit

--------
```

An alternative would be to make it an export template on Device, and then iterate over the interfaces on each device.

If you can turn it into something useful, you might want to submit a completed version of this to the [customizations](https://github.com/netbox-community/customizations) repo.

**Reply by Prestwickuk** [reply]
Thank you very much, I'll use this and see where it takes me! :)
